### PR TITLE
Detect dynamic register keywords 4683 v1

### PR DESCRIPTION
**Pull request by catenacyber**
Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/4683

Describe changes:
- detect: helper to have pure rust keywords
- make keywords registration dynamic
- detect/snmp: move keywords to rust
- convert unit test DetectSNMPCommunityTest to a SV test.
- snmp.pdu_type use a generic uint32 for detection, allowing `>2` and such

SV_BRANCH=https://github.com/OISF/suricata-verify/pull/1804

Continuation of https://github.com/OISF/suricata/pull/9871 after merge of https://github.com/OISF/suricata/pull/10819

After the merge of loggers, pure rust plugins will need pure rust keywords.
The plan is to do this for all rust app-layers, now only done for SNMP, which has both integers and buffers as keywords.

### CODE REVIEW

**Comment by suricata-qa** [comment]
WARNING:
 
 | field | baseline | test | % |
|--|--|--|--| 
| | build_asan | | |
 

Pipeline 20365

**Comment by suricata-qa** [comment]
WARNING:
 
 | field | baseline | test | % |
|--|--|--|--| 
| | build_asan | | |
 

Pipeline 20366

**Comment by suricata-qa** [comment]
WARNING:
 
 | field | baseline | test | % |
|--|--|--|--| 
| | build_asan | | |
 

Pipeline 20367

**Comment by victorjulien** [comment]
Set to draft to avoid the force pushes monopolizing qalab runtime.

File: rust/src/snmp/detect.rs [hunk 1]
```diff
@@ -17,42 +17,263 @@
...
+
+static mut G_SNMP_VERSION_KW_ID: c_int = 0;
+static mut G_SNMP_VERSION_BUFFER_ID: c_int = 0;
+static mut G_SNMP_PDUTYPE_KW_ID: c_int = 0;
+static mut G_SNMP_PDUTYPE_BUFFER_ID: c_int = 0;
+static mut G_SNMP_USM_BUFFER_ID: c_int = 0;
+static mut G_SNMP_COMMUNITY_BUFFER_ID: c_int = 0;
+
+#[no_mangle]
+unsafe extern "C" fn snmp_detect_version_setup(
```
Comment (jasonish) [inline rust/src/snmp/detect.rs:39]: naming: no_mangle and extern C should follow C naming conventions.

Comment (jasonish) [reply rust/src/snmp/detect.rs:39]: As this isn't `pub`, and just a function pointer, you could drop the `no_mangle` and use Rust conventions. Meaning you could drop the prefixes in the function name as well, as the visibility is just this module.

Comment (catenacyber) [reply rust/src/snmp/detect.rs:39]: What does the `no_mangle` do ?
I thought I needed it for the ABI

Comment (jasonish) [reply rust/src/snmp/detect.rs:39]: No mangle just keeps the name as-is in the symbol table so linkers can find it.  If you just point to it using a function pointer its not needed and best let Rust do its name mangling.

The `extern` is for the ABI.

Comment (catenacyber) [reply rust/src/snmp/detect.rs:39]: Thanks Jason, makes sense

**Comment by suricata-qa** [comment]
WARNING:
 
 | field | baseline | test | % |
|--|--|--|--| 
| | build_asan | | |
 

Pipeline 20385

**Comment by catenacyber** [comment]
So, problem statement.
I want to have C functions that are only called by rust code.
(Workaround would be to add usages in C)
Linker is not happy about it as it optimizes, and removes unused code, just to say afterwards that there are `undefined reference`

I tried in code
- attribute used
- attribute externally_visible
in compilation
- Did not manage to use `-Wl,--start-group` with autotool Makefile.am (mixing of ldflags and ldadd)
- Double libc.a librust.a libc.a librust.a, so that each definition comes after a use (also just libc.a librust.a libc.a)
- Tried to mix both archives into a single one, but then this single archive gets optimized and main.c ends up having unreferenced stuff

By the way, I do not understand how unused `ConfGetDouble` gets linked in the end in suricata

**Comment by jasonish** [comment]
> So, problem statement. I want to have C functions that are only called by rust code. (Workaround would be to add usages in C) Linker is not happy about it as it optimizes, and removes unused code, just to say afterwards that there are `undefined reference`
> 
> I tried in code :
> 
>     * attribute used
> 
>     * attribute externally_visible
> 
> 
> in compilation :
> 
>     * Did not manage to use `-Wl,--start-group` with autotool Makefile.am (mixing of ldflags and ldadd)
> 
>     * Double libc.a librust.a libc.a librust.a, so that each definition comes after a use (also just libc.a librust.a libc.a)
> 
>     * Tried to mix both archives into a single one, but then this single archive gets optimized and main.c ends up having unreferenced stuff
> 
> 
> By the way, I do not understand how unused `ConfGetDouble` gets linked in the end in suricata

I am not sure why the simple library example is not doing the following:
```
"-Wl,--start-group,../../../src/libsuricata_c.a,../../../rust/target/debug/libsuricata_rust.a,--end-group"
```
it seems to happen automatically for `suricata`, and the Makefiles seem similar in this respect, but `libtool` isn't doing its thing...  For the external library user build we'll have to replicate that in `libsuricata-config`.

I wouldn't add `--start-group`, etc manually.  In `master`, that is what libtool is doing. So we just need to coerce to do it more places.  As for `libsuricata-config`, I can look at that in the next few days.

**Comment by jasonish** [comment]
> So, problem statement. I want to have C functions that are only called by rust code. (Workaround would be to add usages in C) Linker is not happy about it as it optimizes, and removes unused code, just to say afterwards that there are `undefined reference`
> 
> I tried in code :
> 
>     * attribute used
> 
>     * attribute externally_visible
> 
> 
> in compilation :
> 
>     * Did not manage to use `-Wl,--start-group` with autotool Makefile.am (mixing of ldflags and ldadd)
> 
>     * Double libc.a librust.a libc.a librust.a, so that each definition comes after a use (also just libc.a librust.a libc.a)
> 
>     * Tried to mix both archives into a single one, but then this single archive gets optimized and main.c ends up having unreferenced stuff
> 
> 
> By the way, I do not understand how unused `ConfGetDouble` gets linked in the end in suricata

I don't think the linker is removing the code, `-rdynamic` is supposed to make sure that doesn't happen. Its just the link only looks at libraries listed after to resolve patterns.  So when we do C library first, then rust, it can't find it.

So portable doing the start-group type stuff is probably the way to resolve this.

**Comment by jasonish** [comment]
One thought... Can you move those functions over to Rust, even if very C like, then just call more primitives on the C that are for sure used?

**Comment by jasonish** [comment]
> Did not manage to use `-Wl,--start-group` with autotool Makefile.am (mixing of ldflags and ldadd)

If I recall correctly, `libtool` will re-order all the compilation args in such away that this doesn't work unfortunately.

**Comment by catenacyber** [comment]
> > One thought... Can you move those functions over to Rust, even if very C like, then just call more primitives on the C that are for sure used?
> 
> Good idea

But does not seem possible for `DetectHelperKeywordRegister` at least...

**Comment by catenacyber** [comment]
> I don't think the linker is removing the code, -rdynamic is supposed to make sure that doesn't happen. Its just the link only looks at libraries listed after to resolve patterns. So when we do C library first, then rust, it can't find it.

> So portable doing the start-group type stuff is probably the way to resolve this.

Yes, but that does not look so portable...

**Comment by catenacyber** [comment]
Continued in https://github.com/OISF/suricata/pull/10992

**Comment by jasonish** [comment]
> > I don't think the linker is removing the code, -rdynamic is supposed to make sure that doesn't happen. Its just the link only looks at libraries listed after to resolve patterns. So when we do C library first, then rust, it can't find it.
> 
> > So portable doing the start-group type stuff is probably the way to resolve this.
> 
> Yes, but that does not look so portable...

Possibly.  It is required for plugins, but we only test that on Linux.  I know they're broken on Windows. Possibly Mac as well.

**Comment by jasonish** [comment]
> > > One thought... Can you move those functions over to Rust, even if very C like, then just call more primitives on the C that are for sure used?
> > 
> > 
> > Good idea
> 
> But does not seem possible for `DetectHelperKeywordRegister` at least...

Function point in the context object?

**Comment by catenacyber** [comment]
> > But does not seem possible for `DetectHelperKeywordRegister` at least...
> 
> Function point in the context object?

Possible, but I am not a big fan of this context object... I see that for frames, you have set up empty stubs for cfg test :-)

**Comment by jasonish** [comment]
> > > But does not seem possible for `DetectHelperKeywordRegister` at least...
> > 
> > 
> > Function point in the context object?
> 
> Possible, but I am not a big fan of this context object... I see that for frames, you have set up empty stubs for cfg test :-)

Is anyone?

I've prototyped re-working the build to this order:
- generate headers from rust code
- build C code into library (no main)
- build rust code into library that also contains the C library code, so a single library
- build suricata application from single library

It works, and makes the C code available to Rust unit tests.  Essentially eliminating the need for the context object.  Doesn't work with ASAN -- builds fail. So a bit unfortunate on that front.

**Comment by catenacyber** [comment]
Looks nice.

How does it fail with ASAN ?

**Comment by jasonish** [comment]
> How does it fail with ASAN ?

At link time..  Scenario is:
- build C objects files with asan
- build rust library
- can build suricata application
- cannot build rust unit tests, they're essentially an application

Built the Rust "test" apps require manual linking with `asan`, however when manually linking with `asan` it must be the first lib linked, and I can't make it the first with Rust flags and what not.  Tried using the Rust -Z sanitize stuff, but no luck there either.

**Comment by catenacyber** [comment]
In any case, I think that the functionality of having C functions only called by rust could be useful.

But if it gets too painful for this PR, I will resort to use them in C also...
But it looks not far in next version...
MacOS build is failing, but I think I found a "bug" in the build system for fuzz targets

**Comment by jasonish** [comment]
> In any case, I think that the functionality of having C functions only called by rust could be useful.
> 
> But if it gets too painful for this PR, I will resort to use them in C also... But it looks not far in next version... MacOS build is failing, but I think I found a "bug" in the build system for fuzz targets

if `-lsuricata_c -lsuricata_rust -lsuricata_c` does allow it to work, its a reasonable work-around as long as we're stuck with libtool.  I'll look at getting rid of libtool after we remove the option of bundled/unbundled libhtp.

**Pull request closed without merging**